spec/lang/machine.md

```diff
@@ -48,17 +48,15 @@ struct StackFrame<M: Memory> {
 
     /// `next_block` and `next_stmt` describe the next statement/terminator to execute (the "program counter").
     /// `next_block` identifies the basic block,
-    next_block: BbName,
+    /// next_block is None after calling a function without giving a basic block to return to. UB is raised after the function returns.
+    next_block: Option<BbName>,
```

Comment:
I don't think we want to allow this to be `None`. Why shouldn't we have UB immediately any time this would be set to `None`?

Reply:
In the syntax for Terminator::Call, both the return place and the next block are optional (separately). This is set to None in any Call (or CallIntrinsic) without a next block, and that is sensible when calling a non-returning function. Returning anyway will then be UB, of course.

The conceptual idea with the field being Option would be that a stack frame may keep existing even when it will never execute another statement or terminator.

Reply:
> a stack frame may keep existing even when it will never execute another statement or terminator.

I don't think that is something we want though.
Thinking about `become`, I would expect the old stack frame to be popped and then the new one to be pushed.

Reply:
Sure, what about that? We can already represent them. Just make a call that has neither a return block nor a return place.

Reply:
A call to such a function is exactly the situation where (with this patch) next_block is set to None, as in "the caller will not be asked to execute another step, so it doesn't need to define what that step would be". Before this patch, the callee's caller_return_info.next_block would instead be set to None, so that if and when it tries to return, that will be UB.

With this patch, that return would technically be fine, but the next step in that thread is UB. This partially unifies the behaviour with returning from the bottom frame, which is not UB even if the function doesn't have a return local, although I would also like to change that towards requiring a return local to return at all.

Reply:
>Okay. I just don't think that's a good change though -- the current behavior seems better to me. I'd like updating the "current basic block" in the caller's frame to happen upon return, not during the Call. It helps with diagnostics, and it IMO better represents the intended semantics: the jump happens on Return.

You have convinced me of that now. The rationale of making this change was more that the Return in one function shouldn't need to know anything about the caller's structure or stack frame. I previously considered Call to be something that is evaluated and then its done, and the entirety of the callee happens afterwards.

However, together with what I've learned from https://github.com/rust-lang/unsafe-code-guidelines/issues/417, I'm now thinking that maybe Call should actually be a two-step thing: Pre-call, it prepares the return place (e.g. either allocate new place, or just pass the actual destination) and evaluates the arguments to form the callee's StackFrame. Post-call, if the return place was a temporary, copy to the actual destination and deallocate, and finally update next_block.

In other words, while a stack frame is not at the top of the stack, it has a Call "on hold" that will be finished after the function above it evaluates a Return. I'll have to see how it would end up looking in practice.

Reply:
Yes, that is a way to think about this. The `CallerReturnInfo` is basically the environment of a caller-defined closure that will be run after the callee returns to run the post-call actions. We don't actually want this to be a closure since having the data explicit will make it a lot easier to formally reason about this specification. (With such reasoning, equality might come in, and a struct like `CallerReturnInfo` has a much clearer notion of equality than a closure.)